packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3613,9 +3642,152 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+    return Rect.fromLTWH(view.padding.left / view.devicePixelRatio, view.padding.top / view.devicePixelRatio, visibleScreenSize.width, visibleScreenSize.height);
+  }
+
+  void _handleContextMenuOnScroll(ScrollNotification notification) {
+    if (_webContextMenuEnabled) {
+      return;
+    }
+    if (!_platformSupportsFadeOnScroll) {
+      _selectionOverlay?.updateForScroll();
+      return;
+    }
+    // When the scroll begins and the toolbar is visible, hide it
+    // until scrolling ends.
+    //
+    // The selection and renderEditable need to be visible within the current
+    // viewport for the toolbar to show when scrolling ends. If they are not
+    // then the toolbar is shown when they are scrolled back into view, unless
+    // invalidated by a change in TextEditingValue.
+    if (notification is ScrollStartNotification) {
+      if (_valueWhenToolbarShowScheduled != null) {
+        return;
+      }
+      final bool toolbarIsVisible = _selectionOverlay != null
+                                  && _selectionOverlay!.toolbarIsVisible
+                                  && !_selectionOverlay!.spellCheckToolbarIsVisible;
+      _valueWhenToolbarShowScheduled = toolbarIsVisible ? _value : null;
+      if (_valueWhenToolbarShowScheduled != null) {
+        _selectionOverlay?.hideToolbar();
+      }
+    } else if (notification is ScrollEndNotification) {
+      if (_valueWhenToolbarShowScheduled == null) {
+        return;
+      }
+      if (_valueWhenToolbarShowScheduled != _value) {
+        // Value has changed so we should invalidate any toolbar scheduling.
+        _valueWhenToolbarShowScheduled = null;
+        _disposeScrollNotificationObserver();
+        return;
+      }
+
+      final Rect renderEditableBounds = MatrixUtils.transformRect(renderEditable.getTransformTo(null), renderEditable.paintBounds);
+      final Rect viewportRect = _calculateViewportRect();
+      final bool selectionIsVisible = renderEditable.selectionStartInViewport.value || renderEditable.selectionEndInViewport.value;
+      final bool? renderEditableInViewport = _renderEditableInViewport(notification);
+      final bool renderEditableIsVisible = renderEditableInViewport == null || (renderEditableInViewport && viewportRect.overlaps(renderEditableBounds));
+
+      if (selectionIsVisible && renderEditableIsVisible) {
+        showToolbar();
+        _valueWhenToolbarShowScheduled = null;
+      }
+    }
+  }
+
+  bool? _renderEditableInViewport(ScrollNotification notification) {
+    RenderAbstractViewport? closestViewport = RenderAbstractViewport.maybeOf(renderEditable);
+    if (closestViewport == null) {
+      return null;
+    }
+    while (closestViewport != null) {
+      final RevealedOffset leadingEdgeOffset = closestViewport.getOffsetToReveal(renderEditable, 0.0);
+      final RevealedOffset trailingEdgeOffset = closestViewport.getOffsetToReveal(renderEditable, 1.0);
+      final double currentOffset = notification.metrics.pixels;
```

Comment:
This assumes the notification comes from the closest viewport?

Reply:
Good point, I have addressed this in https://github.com/flutter/flutter/pull/138313/commits/cae38b72b18848f837c43921eb7832a62a261223.

Reply:
I landed on this approach https://github.com/flutter/flutter/pull/138313/commits/599beb3cf7d5ad1c3101740ea9aa2648018ff90d which checks if the `renderEditable.paintBounds` overlaps with the `viewport.paintBounds`, my previous approach I linked would hide the context menu when the `renderEditable` was even partially occluded. I verified this approach covers the cases we discussed.